### ISSUE TITLE
Create subroutine to map commodities based on hs6->fcl one-to-one links

### ISSUE THREAD

**Issue by malexan** [SWS-Methodology/faoswsTrade#122]
It is required to solve #78 and generally improve mapping accuracy. 

In cases when we have one-to-one links from HS codes of length 6 to FCL codes, we do not need detailed HS codes.

Take into account:
1. There are cases with one-to-many HS->FCL links. All these cases should be proceeded with usual approach.
2. There are cases where a country reports HS codes shorter than 6. They also should be proceeded usual way. Additionally we could check do short HS codes have one-to-one links with FCL and use new approach.

The algorithm: 
1. Create in HS->FCL map additional column with HS code of length not longer than 6.
    * Create a numeric vector from `fromcode` to `tocode`
    * Duplicate records for each number from the vector
2. Mark all reporter/flow/hs6 combinations what have one-to-one link with FCL. 
3. Create in trade data column with HS code of length not longer than 6.
4. Mark in trade data all records suitable for mapping based on HS6.
5. Map all suitable records using new approach.
6. Map other records using old approach.

From the perspective of #94 the procedure should not be hidden inside of `mapHS2FCL` function.

**Comment by malexan** [comment]
We have to convert every records `fromcode`-`tocode` range into a multiple records with each unique fcl code.

```r
hsfclmap %>% 
  mutate(hs6 = str_sub(fromcode, end = 6L),
         hs6to = str_sub(tocode, end = 6L),
         noteq = hs6 != hs6to) %>% 
  summarise(count = sum(noteq),
            prop = count / n())
```  
```
# A tibble: 1 × 2
  count      prop
  <int>     <dbl>
1 68143 0.1647243
```

**Comment by malexan** [comment]
Moving from ranges to unique hs codes is very slow operation due to massive building and binding of data frames.

```r
hsfclmap1 <- sample_n(hsfclmap, 10^5)

microbenchmark::microbenchmark(
  extract_hs6fclmap(hsfclmap1),
  extract_hs6fclmap(hsfclmap1, parallel = TRUE),
  times = 1L
)
```
```
Unit: seconds
                                          expr      min       lq     mean   median       uq
                  extract_hs6fclmap(hsfclmap1) 132.9220 132.9220 132.9220 132.9220 132.9220
 extract_hs6fclmap(hsfclmap1, parallel = TRUE)  93.8749  93.8749  93.8749  93.8749  93.8749
      max neval
 132.9220     1
  93.8749     1
```

**Comment by malexan** [comment]
Closed in branch [iss122hs6map](https://github.com/SWS-Methodology/faoswsTrade/tree/iss122hs6map).

**Comment by chrmongeau** [comment]
@malexan I see that in `mapHS6toFCL` the [mapping is done reporter-specific](https://github.com/SWS-Methodology/faoswsTrade/blob/02c6d6761ac72fac3a7894af9324fe43b1f6205c/R/mapHS6toFCL.R#L17). Have you considered to remove the reporter (even the flow?) in the mapping? The 6-digit HS should be common to all reporters, so if for some of them there is no mapping for that specific year, the use of the common hs6 should improve the link creation.

**Comment by malexan** [comment]
Yep, @chrMongeau, I tried without flow and without reporter. But in this case amount of one-to-one hs6->fcl links declines. 

I am thinking about a method for unmapped HS codes. We can use information from multi-matched hs6->fcl links to increase accuracy. For example, if 101010 is linked to 123 for 3 years and to 124 for 4 years, we can assume that in unmapped case 101010 goes to 124 (3 < 4). And information about specific reporter and flow will increase accuracy.

But firstly let's check results of current HS6 approach. Could you run the module from branch `issue122hs6map` to check does it work in real environment and to compare results?

**Comment by chrmongeau** [comment]
> But firstly let's check results of current HS6 approach. Could you run the module from branch issue122hs6map to check does it work in real environment and to compare results?

@malexan Done. It worked fine and the results seem very good. I sent the report via email.

I had a look at `tldata_hsfcl_nolinks.csv` (year = 2014) and there are some countries that present a high number of unmapped codes, e.g., Aruba and Sierra Leone (by the way, Canada is not there: all problems were resolved?). I had a look at Sierra Leone and found that the mapping that is being used by the HS>FCL mapping is made of 22 rows:

```
> filter(hsfclmap, area == 197)
# A tibble: 22 × 6
    area  flow fromcode   tocode   fcl recordnumb
   <int> <int>    <chr>    <chr> <int>      <int>
1    197     1  2042200  2042200  1035    5922875
2    197     1  2072700  2072700  1080    5922876
3    197     1  2073200  2073200  1069    5922877
4    197     1  2109910  2109910  1164    5922878
5    197     1  2109990  2109990  1164    5922879
6    197     1  4069900  4069900   901    5922880
7    197     1  4070050  4070050  1062    5922881
8    197     1  9061100  9061100   693    5922882
9    197     1 11032000 11032000   111    5922883
10   197     1 12072010 12072010   329    5922884
# ... with 12 more rows
```

that map 15 different HS codes (I'm not considering the ranges nor the flows, bear with me):

```
> filter(hsfclmap, area == 197)  %>% count(fromcode)
# A tibble: 15 × 2
   fromcode     n
      <chr> <int>
1  11032000     2
2  12072010     2
3  12079950     1
4  20097100     2
5  20097900     1
6  20098020     1
7   2042200     1
8   2072700     2
9   2073200     2
10 21061030     1
11  2109910     2
12  2109990     1
13  4069900     1
14  4070050     2
15  9061100     1
```

Now, the original `hsfclmap3` maps 2,225 HS codes:

```
> filter(hsfclmap3, area == 197)  %>% count(fromcode)
# A tibble: 2,225 × 2
   fromcode     n
      <chr> <int>
1  01011100     2
2  01011190     2
3  01011990     2
4  01012000     2
5  01021000     2
6  01021010     2
7  01021020     2
8  01021090     2
9  01029005     2
10 01029010     2
# ... with 2,215 more rows
```

The `hsfclmap` is made up of codes that appears after 2014, while in the original `hsfclmap3` map there were codes (the majority) available before:

```
> filter(hsfclmap3, area == 197) %>% count(startyear, endyear)
Source: local data frame [2 x 3]
Groups: startyear [?]

  startyear endyear     n
      <int>   <int> <int>
1      2002    2002  4424
2      2014    2050    22
```

Probably a code that is not overwritten by a newer mapping should survive even if `endyear` is less than the current year? I did a test with the 443 unmapped codes from Sierra Leone and for 392 of them there is a mapping if one keep the startyear = 2002 & endyear = 2002 mapping.

What do you think?

**Comment by chrmongeau** [comment]
Oh, I think the situation is more or less the same if one applies the same reasoning to Aruba:

Map used in the mapping:

```
> filter(hsfclmap, area == 22)
# A tibble: 17 × 6
    area  flow fromcode   tocode   fcl recordnumb
   <int> <int>    <chr>    <chr> <int>      <int>
1     22     1 12122100 12122100  1293    5922681
2     22     1 12122900 12122900  1293    5922682
3     22     1 15021000 15021000  1225    5922683
4     22     1 15060090 15060090  1129    5922684
5     22     1 15019000 15019000  1066    5922685
6     22     1 15011000 15011000  1043    5922686
7     22     1 15029000 15029000   869    5922687
8     22     1 20089790 20089790   623    5922688
9     22     1 20098100 20098100   622    5922689
10    22     1 20098900 20098900   466    5922690
11    22     1  7099900  7099900   463    5922691
12    22     1  7099910  7099910   463    5922692
13    22     1  7099990  7099990   463    5922693
14    22     2  7099900  7099900   463    5922694
15    22     1  7099300  7099300   394    5922695
16    22     1  7099100  7099100   366    5922696
17    22     1  7099200  7099200   260    5922697
```

Original mapping:

```
> filter(hsfclmap3, area == 22) %>% count(startyear, endyear)
Source: local data frame [10 x 3]
Groups: startyear [?]

   startyear endyear     n
       <int>   <int> <int>
1       1999    1999     2
2       1999    2006     4
3       1999    2011  2080
4       2000    2011     8
5       2001    2011     2
6       2007    2011     4
7       2009    2011   434
8       2010    2011    36
9       2011    2011    32
10      2014    2050    17
```

**Comment by malexan** [comment]
Yep, as about Aruba etc you are describing #77

**Comment by chrmongeau** [comment]
I just checked the ratio of values for which `endyear` is not 2050 and got these results:

```
> hsfclmap3 %>%
	group_by(area) %>%
	summarise(ratio=sum(endyear==2050)/n()) %>%
	do(data.frame(q=quantile(.$ratio, prob=1:10/10)))

               q
10%  0.000000000
20%  0.003451432
30%  0.272509686
40%  0.380030981
50%  0.519637462
60%  0.643911311
70%  0.937415393
80%  0.973674541
90%  0.989341926
100% 1.000000000
```

**Comment by chrmongeau** [comment]
> Yep, as about Aruba etc you are describing #77

OK, I'll leave a comment there.

**Comment by chrmongeau** [comment]
Mmmh, sorry [I indeed commented there](https://github.com/SWS-Methodology/faoswsTrade/issues/77#issuecomment-294152329), but it's better placed here. I'm copy-pasting:

> > If we don't take into account startyear/endyear in hs->fcl mapping table the number of not-mapped HS codes drops significantly
> 
> At least `startyear`, should be taken into account. For instace, in:
> 
> ```
> > filter(hsfclmap3, area == 22, (startyear == 1999 & endyear == 2006) | (startyear == 2007 & endyear == 2011))
> # A tibble: 8 × 8
>    area  flow fromcode   tocode   fcl startyear endyear recordnumb
>   <int> <int>    <chr>    <chr> <int>     <int>   <int>      <int>
> 1    22     1 19010000 19010000    50      1999    2006    5758519
> 2    22     1 19010000 19010000   109      2007    2011    5742740
> 3    22     1 19019010 19019010    50      1999    2006    5758867
> 4    22     1 19019010 19019010   109      2007    2011    5742741
> 5    22     2 19010000 19010000    50      1999    2006    5757429
> 6    22     2 19010000 19010000   109      2007    2011    5741694
> 7    22     2 19019010 19019010    50      1999    2006    5757777
> 8    22     2 19019010 19019010   109      2007    2011    5741693
> ```
> 
> It seems that HS=19010000 (the same for 19019010) was mapped to FCL=50 from 1999 to 2006 and from 2007 until 2011 it was mapped to 109. In the current mapping we don't have these codes:
> 
> ```
> > filter(hsfclmap, area == 22, fromcode %in% c('19010000', '19019010'))
> # A tibble: 0 × 6
> # ... with 6 variables: area <int>, flow <int>, fromcode <chr>, tocode <chr>, fcl <int>, recordnumb <int>
> ```
> 
> The problem is that given that the maximum `endyear` is 2011, these codes are discarded. I think `endyear` should always be expanded until 2050 unless we know that they were changed. For instance, in the previous case, if we had *only*:
> 
> ```
>    area  flow fromcode   tocode   fcl startyear endyear recordnumb
> 1    22     1 19010000 19010000    50      1999    2006    5758519
> ```
> 
> It should become:
> 
> ```
>    area  flow fromcode   tocode   fcl startyear endyear recordnumb
> 1    22     1 19010000 19010000    50      1999    2050    5758519
> ```
> 
> (2006 was changed to 2050). Now, given that we have:
> 
> ```
>    area  flow fromcode   tocode   fcl startyear endyear recordnumb
>   <int> <int>    <chr>    <chr> <int>     <int>   <int>      <int>
> 1    22     1 19010000 19010000    50      1999    2006    5758519
> 2    22     1 19010000 19010000   109      2007    2011    5742740
> ```
> 
> It should become:
> 
> ```
>    area  flow fromcode   tocode   fcl startyear endyear recordnumb
>   <int> <int>    <chr>    <chr> <int>     <int>   <int>      <int>
> 1    22     1 19010000 19010000    50      1999    2006    5758519
> 2    22     1 19010000 19010000   109      2007    2050    5742740
> ```
> 
> (notice that 2011 in the second row was changed to 2050).

**Comment by chrmongeau** [comment]
Let's consider another example. In 2010 country 16 (Bangladesh) is in the `tldata_not_area_in_fcl_mapping` list, given that there it's not present in `hsfclmap` as the maximum `endyear` is 2007, though it would be present if `endyear` is extended to 2050:

```
> year
[1] 2010
> tldata_not_area_in_fcl_mapping %>% distinct(reporter)
# A tibble: 18 × 1
   reporter
      <int>
1         6
2        16
3       252
4        28
5        45
6       299
7        85
8        95
9       103
10      122
11      124
12      133
13      147
14      180
15      188
16      193
17      237
18      225
> hsfclmap %>% filter(area == 16)
# A tibble: 0 × 6
# ... with 6 variables: area <int>, flow <int>, fromcode <chr>, tocode <chr>, fcl <int>,
#   recordnumb <int>
> hsfclmap3 %>% filter(area == 16) %>% count(startyear, endyear)
Source: local data frame [9 x 3]
Groups: startyear [?]

  startyear endyear     n
      <int>   <int> <int>
1      2004    2004     6
2      2004    2005     2
3      2004    2006     6
4      2004    2007  3690
5      2005    2005    26
6      2005    2007    12
7      2006    2006     4
8      2006    2007     4
9      2007    2007    40
```

I'd even apply the same logic to previous years (i.e., extend backwards), with some caution. For instance, Bangladesh is also in `tldata_not_area_in_fcl_mapping` for 2003, even if we have data for it:

```
> year
[1] 2003
> tldata %>% filter(reporter == 16) %>% count(reporter)
# A tibble: 1 × 2
  reporter     n
     <int> <int>
1       16  7934
> hsfclmap %>% filter(area == 16)
# A tibble: 0 × 6
# ... with 6 variables: area <int>, flow <int>, fromcode <chr>, tocode <chr>, fcl <int>, recordnumb <int>
> hsfclmap3 %>% filter(area == 16) %>% count(startyear, endyear)
Source: local data frame [9 x 3]
Groups: startyear [?]

  startyear endyear     n
      <int>   <int> <int>
1      2004    2004     6
2      2004    2005     2
3      2004    2006     6
4      2004    2007  3690
5      2005    2005    26
6      2005    2007    12
7      2006    2006     4
8      2006    2007     4
9      2007    2007    40
```

**Comment by malexan** [comment]
Test runs on 2013 and 2014 show strange outcome: the HS6 mapping algorithm improves the situation with TL data but degrades quality of ES. Changes in number of trade flows with unmapped HS codes from the traditional approach to the HS6 approach:

* TL 2014: from 17885 to 8041;
* TL 2013: from 3363 to 1583;
* ES 2014: from 173 to 4922;
* ES 2013: from 38 to 4866.

As per the design the HS6 approach can't produce results worse than the traditional approach. So we have an implementation error and this branch can't be merged into master now.

**Comment by malexan** [comment]
That was a though hunt!

Now we have when adding mapping step with hs6 (in brackets there is a value in the result of the bug):

* TL 2014: from 17885 to 7565 (was 8041);
* TL 2013: from 3363 to 684 (was 1583);
* ES 2014: from 173 to 54 (was 4922);
* ES 2013: from 38 to 7 (was 4866).

The problem concerned identifiers of HS->FCL mapping table records. Each record has its own unique `recordnumb`. When I extracted the hs6 level mapping table, I preserved this ID and then used it to drop correspondence links from the original mapping table (hs+) what were to be replaced by one-to-one hs6->fcl links.

During the hs6 extraction one ID from original mapping table could be used in several records of HS6 mapping table. And some of HS6->FCL records could be one-to-one, but others - one-to-many. I didn't recognize it and used IDs to drop unnecessary HS+ correspondence records from the original mapping table to increase the speed of mapping process.

So if one original ID covered several HS6 mapping records and among them there were as one-to-one, so one-to-many HS6 links, then such record was incorrectly removed from the original mapping table.
